Fix part of #135: UI Structure:Topic player multiple tabs

nikitamarysolomanpvt opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
fix #135 


## Mock
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/0fad68f5-5ca4-4947-b8db-b4cf07235fd7/Topic-w-o-Play-19
To test this app, follow these steps:
1. In `AndroidManifest` file make `TopicActivity` as launcher activity.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

nikitamarysolomanpvt committed:
multiple tabs

nikitamarysolomanpvt committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

# Conflicts:
#	app/build.gradle
#	app/src/main/AndroidManifest.xml
#	app/src/main/res/values/styles.xml

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt commented:
@BenHenning Please confirm the approach. Please do suggest the test case for multiple tabs.

rt4914 commented:
@nikitamarysolomanpvt 
Revise title:

`Fix #135 :UI Structure:Topic player multiple tabs`

rt4914 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -14,9 +14,8 @@
-    <activity android:name="org.oppia.app.home.HomeActivity">
+    <activity android:name=".topic.TopicActivity"
+              android:theme="@style/OppiaTheme.NoActionBar">
       <intent-filter>
```
`HomeActivity` should be the launcher activity and not `TopicActivity` because if this PR gets merged than again we will have to make one more PR to make launcher activity back to `HomeActivity` which would be incorrect.

I suggest in such cases keep the natural flow of apps intact and in PR description write steps on how to check/run this PR.

BenHenning replied:
+1 though feel free to add a button or something in the home fragment to easily route to the topic activity to make things easier.

nikitamarysolomanpvt replied:
resolved

rt4914 commented on app/build.gradle:
```diff
@@ -62,6 +62,7 @@ dependencies {
       'androidx.constraintlayout:constraintlayout:1.1.3',
       'androidx.core:core-ktx:1.0.2',
       'androidx.lifecycle:lifecycle-runtime-ktx:2.2.0-alpha03',
+      'com.google.android.material:material:1.0.0',
```
Maintain the alphabetical order while adding libraries

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -4,20 +4,80 @@ import android.view.LayoutInflater
+    adapter.addFragment(TopicOverviewFragment(), "OVERVIEW")
+    adapter.addFragment(TopicPlayFragment(), "PLAY")
+    adapter.addFragment(TopicTrainFragment(), "TRAIN")
+    adapter.addFragment(TopicReviewFragment(), "REVIEW")
```
Prefer using this titles from `strings.xml` so that in future if we translate this app to other language that would be easy.

nikitamarysolomanpvt replied:
Resolved

rt4914 reviewed (commented):
LGTM.
Left some nit changes and add test-cases.

Also @BenHenning PTAL, basically can you verify this approach first so that @nikitamarysolomanpvt can start working on test-cases for this PR.

nikitamarysolomanpvt committed:
cust_args

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -4,20 +4,80 @@ import android.view.LayoutInflater
+  private val tabIcons =
+    intArrayOf(
+      R.drawable.ic_overview_white_24dp,
+      R.drawable.ic_overview_white_24dp,
```
Shouldn't we be using the other icons for these, instead?

nikitamarysolomanpvt replied:
Ya i did't get icons so once i get them will replace this

veena14cs replied:
You can add TODO to replace these icons.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -4,20 +4,80 @@ import android.view.LayoutInflater
+      R.drawable.ic_overview_white_24dp
+    )
+
+  private fun setupTabIcons() {
```
Nit: ``setUpTabIcons`` ('setup' is a noun whereas 'set up' is a verb, a method names should be verbs or verb phrases).

nikitamarysolomanpvt replied:
resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -4,20 +4,80 @@ import android.view.LayoutInflater
+    viewPager.adapter = adapter
+  }
+
+  internal inner class ViewPagerAdapter(manager: FragmentManager) : FragmentPagerAdapter(manager) {
```
Can this be private, instead of internal?

nikitamarysolomanpvt replied:
resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -4,20 +4,80 @@ import android.view.LayoutInflater
+  internal inner class ViewPagerAdapter(manager: FragmentManager) : FragmentPagerAdapter(manager) {
+    private val mFragmentList = ArrayList<Fragment>()
+    private val mFragmentTitleList = ArrayList<String>()
+    override fun getItem(position: Int): Fragment {
```
Nit: newline before this method for consistency.

nikitamarysolomanpvt replied:
resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -4,20 +4,80 @@ import android.view.LayoutInflater
+  }
+
+  internal inner class ViewPagerAdapter(manager: FragmentManager) : FragmentPagerAdapter(manager) {
+    private val mFragmentList = ArrayList<Fragment>()
```
Given the tabs are all known ahead of time, could we keep these lists as immutable and pass in the fragments/titles when constructing the adapter?

nikitamarysolomanpvt replied:
updated the code

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -4,20 +4,80 @@ import android.view.LayoutInflater
+
+  internal inner class ViewPagerAdapter(manager: FragmentManager) : FragmentPagerAdapter(manager) {
+    private val mFragmentList = ArrayList<Fragment>()
+    private val mFragmentTitleList = ArrayList<String>()
```
Nit here & above: please omit the ``m`` for fields. We don't use the AOSP style of member naming. See: https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -4,20 +4,80 @@ import android.view.LayoutInflater
     return binding.root
   }
+
+  private fun setupViewPager(viewPager: ViewPager) {
```
Ditto for this method name.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/res/layout/topic_fragment.xml:
```diff
@@ -1,18 +1,43 @@
+        app:popupTheme="@style/ThemeOverlay.AppCompat.Light"
+        app:title="Topic"/>
+      <com.google.android.material.tabs.TabLayout
+        android:id="@+id/tabs"
```
Perhaps: ``topic_tabs_container``? For this & other new identifiers introduced in this PR, please make them contextual so that it's clear in code what sort of content they may contain.

nikitamarysolomanpvt replied:
Resolved

BenHenning reviewed (changes requested):
Thanks @nikitamarysolomanpvt! Can you add test cases? E.g. it would be nice to see TopicFragment tests for each of the four tabs that:
- Verify the presence of the tab
- Verify navigating to the tab shows the correct content
- Verify that a configuration change keeps the same tab open

And one test for verifying that the initial tab that's open by default is correct for a total of 13 tests.

veena14cs commented on app/src/main/res/values/styles.xml:
```diff
@@ -1,49 +1,42 @@
     <item name="android:windowExitAnimation">@anim/slide_down</item>
   </style>
-
+  <style name="OppiaTheme.NoActionBar">
```
You can use `OppiaThemeWithoutActionBar` which is already defined and add you attributes there.

rt4914 commented:
@nikitamarysolomanpvt Search for TODO(#135) in code which can be fixed in your PR.

nikitamarysolomanpvt committed:
multiple tabs will getting topic id from home
Merge branch 'develop' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
#	domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt

nikitamarysolomanpvt committed:
test cases,multiple tabs will getting topic id from home
Merge branch 'develop' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
#	domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt commented:
Configuration change test case is pending.
If someone worked on...  after switching activity how to update test rule for configuration change ,please do suggest the solution.

rt4914 commented on app/build.gradle:
```diff
@@ -87,6 +88,7 @@ dependencies {
       'androidx.test.ext:junit:1.1.1',
       'androidx.test:runner:1.2.0',
       'com.google.truth:truth:0.43',
+      'androidx.test.espresso:espresso-intents:3.1.0',
```
Optional nit: Maintain alphabetical order.

nikitamarysolomanpvt replied:
Removed

rt4914 commented on app/src/main/java/org/oppia/app/home/RouteToTopicListener.kt:
```diff
@@ -0,0 +1,6 @@
+
+/** Listener for when an activity should route to a exploration. */
+interface RouteToTopicListener {
+  fun routeToTopic(explorationId: String)
```
Change `explorationId` to `topicId` and also change the kDoc.

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicActivity.kt:
```diff
@@ -45,5 +50,11 @@ class TopicActivity : InjectableAppCompatActivity(), RouteToQuestionPlayerListen
 
   companion object {
     internal const val TAG_CONCEPT_CARD_DIALOG = "CONCEPT_CARD_DIALOG"
+    /** Returns a new [Intent] to route to [TopicActivity] for a specified topic ID. */
```
This comment is for line 37 in this file, `routeToTopicPlayFragment ` function, finish this functionality. As mentioned earlier you will need to fix all TODOs corresponding to #135 in current code.

nikitamarysolomanpvt replied:
Will do that in different PR

nikitamarysolomanpvt replied:
Resolved in #285

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicActivityPresenter.kt:
```diff
@@ -1,19 +1,26 @@
 import org.oppia.app.R
 import org.oppia.app.activity.ActivityScope
+import org.oppia.app.player.exploration.EXPLORATION_ACTIVITY_TOPIC_ID_ARGUMENT_KEY
+import org.oppia.app.player.exploration.ExplorationFragment
```
Remove unused imports.

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicFragment.kt:
```diff
@@ -11,13 +11,15 @@ import javax.inject.Inject
 
   override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-    return topicFragmentPresenter.handleCreateView(inflater, container)
+    topicId = arguments!!.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY)
```
There is warning on this line. Mainly `topicId` is non-null and the argument can provide a null value. So type mismatch warning.

nikitamarysolomanpvt replied:
Added a null check and as per our discussion we are avoiding the warning for now.

rt4914 commented on app/src/main/res/values-v21/strings.xml:
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources></resources>
```
I think this file is of no use. If you are keeping it than add EOF.

nikitamarysolomanpvt replied:
Removed

rt4914 commented on domain/build.gradle:
```diff
@@ -43,7 +43,8 @@ dependencies {
       'com.google.dagger:dagger:2.24',
-      "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+      "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version",
+      'com.google.code.gson:gson:2.8.2'
```
`Optional Nit`: Alphabetical order

rt4914 replied:
Also, why do you need domain layer changes in this PR?

nikitamarysolomanpvt replied:
Removed

rt4914 commented on model/src/main/proto/interaction_object.proto:
```diff
@@ -48,4 +48,3 @@ message Fraction {
   int32 numerator = 3;
   int32 denominator = 4;
 }
-
```
undo this change.

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/sharedTest/java/org/oppia/app/topic/TopicActivityTest.kt:
```diff
@@ -25,9 +40,41 @@ import javax.inject.Singleton
-  fun testTopicActivity_loadTopicFragment_hasDummyString() {
-    ActivityScenario.launch(TopicActivity::class.java).use {
-      onView(withId(R.id.dummy_text_view)).check(matches(withText("This is dummy TextView for testing")))
+  fun testTopicActivity_clickDummyButton_showsTopicAcitvityWithMultipleTabs_isTabLayoutDisplayed() {
```
Typo in `Activity`

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/sharedTest/java/org/oppia/app/topic/TopicActivityTest.kt:
```diff
@@ -2,20 +2,35 @@ package org.oppia.app.topic
+import org.oppia.app.recyclerview.RecyclerViewMatcher.Companion.atPosition
 import org.oppia.util.threading.BackgroundDispatcher
 import org.oppia.util.threading.BlockingDispatcher
 import javax.inject.Singleton
```
Remove unused imports.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

rt4914 reviewed (changes requested):
Trying adding following test cases:

`testTopicActivity_swipePage_swipePage_topicTrainTabIsDisplayed`
Similar for other tabs too.

`testTopicActivity_clickDummyButton_opensTopicActivity` in this use you check whether the TopicActivity opens with correct intent or not. Check `TopicTrainFragmentTest` for sample code `testTopicTrainFragment_loadFragment_selectSkills_clickStartButton_skillListTransferSuccessfully`

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt commented:
> @nikitamarysolomanpvt Search for TODO(#135) in code which can be fixed in your PR.

This PR is focusing on only multiple tabs and the navigation to different tabs by swipe or on tab click, as well as navigation from HomeActivity to TopicActivity on button click and passing topic Id to all fragments in side TopicFragment from HomeActivity.
Navigation from the content of different fragments of Topic to another fragment of topic will be implemented in separate PR

veena14cs commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -64,4 +66,9 @@ class HomeFragmentPresenter @Inject constructor(
   }
+
+  fun topicButtonClick(v: View) {
+    logger.d("HomeFragment", "Successfully loaded topic")
```
You can define TAG_HOME_FRAGMENT.

nikitamarysolomanpvt replied:
Resolved

veena14cs commented on app/src/main/java/org/oppia/app/topic/TopicFragment.kt:
```diff
@@ -11,13 +11,15 @@ import javax.inject.Inject
 
   override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-    return topicFragmentPresenter.handleCreateView(inflater, container)
+    topicId =if(arguments!=null) arguments!!.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) else ""
```
Give space after `= if` or use cmd+option+L

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt committed:
separate test cases for all four tabs

nikitamarysolomanpvt commented:
Resolved.
Let me know if you have any suggestion in the test cases implemented.

> Thanks @nikitamarysolomanpvt! Can you add test cases? E.g. it would be nice to see TopicFragment tests for each of the four tabs that:
> 
> * Verify the presence of the tab
> * Verify navigating to the tab shows the correct content
> * Verify that a configuration change keeps the same tab open
> 
> And one test for verifying that the initial tab that's open by default is correct for a total of 13 tests.

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -30,7 +31,8 @@ class HomeFragmentPresenter @Inject constructor(
   private val routeToExplorationListener = activity as RouteToExplorationListener
-
+  private val routeToTopicListener = activity as RouteToTopicListener
+  private val TAG_HOME_FRAGMENT = "HomeFragment"
```
Declare this in similar way `EXPLORATION_ID` has been declared, on top of the class.

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicFragment.kt:
```diff
@@ -10,14 +10,21 @@ import javax.inject.Inject
+    @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+    topicId =
+      if (arguments != null && arguments!!.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) != null) arguments!!.getString(
+        TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY
```
This is creating confusion in brackets:

better way:

`if (arguments != null && arguments!!.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) != null)`
`    arguments!!.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) else ""`

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/sharedTest/java/org/oppia/app/utility/TabMatcher.kt:
```diff
@@ -0,0 +1,50 @@
+import org.hamcrest.Matcher
+
+/** This class mainly provides a custom matcher to test whether the current tab title is correctly selected in TabLayout. */
+class TabMatcher {
```
As informed earlier you were suppose to add the link from where you are using this code. Please add links if you are taking some code snippet from internet.

nikitamarysolomanpvt replied:
Okay

rt4914 commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,347 @@
+    activityTestRule.launchActivity(null)
+    onView(
+      allOf(
+        withText("REVIEW"),
```
Here and everywhere else, get tab-names from `strings.xml`

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,347 @@
+    ).perform(click())
+
+    onView(withId(R.id.topic_tabs_container)).check(matches(matchCurrentTabTitle("OVERVIEW")))
+
```
Remove this extra line.

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,347 @@
+
+  @Test
+  @UiThread
+  fun testTopicActivity_isOverviewTabSelectedAndContentMatched() {
```
This naming is incorrect:

Suggestion: `testTopicActivity_overviewTabIsSelectedWithCorrectContent_isSuccessful`

Not sure if the suggestion is correct but you should not start and action with `is` in `isOverviewTab...` because that is a form of question.

Update test-case names everywhere.

nikitamarysolomanpvt replied:
Renamed

rt4914 reviewed (approved):
Some minor changes and do update test-cases name, currently they are not correct I think.

veena14cs commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,347 @@
+    activityTestRule.launchActivity(null)
+    onView(withId(R.id.topic_tabs_container)).perform(click()).check(matches(isDisplayed()))
+    onView(withId(R.id.topic_tabs_container)).perform(click()).check(matches(isDisplayed()))
+
```
remove extra space.

nikitamarysolomanpvt replied:
Resolved

veena14cs commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,347 @@
+    onView(withText("Master These Skills")).check(matches(isDisplayed()))
+  }
+
+  fun testTopicActivity_clickOnReviewTab_configurationChange_isSameTabAndItsDataDisplayed() {
```
Is this a test case? Its not annotated with `Test`.

nikitamarysolomanpvt replied:
Resolved

veena14cs commented on app/src/sharedTest/java/org/oppia/app/utility/TabMatcher.kt:
```diff
@@ -0,0 +1,50 @@
+      }
+    }
+  }
+
```
Remove extra space.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
merged with develop

nikitamarysolomanpvt committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

nikitamarysolomanpvt committed:
revert back to develop

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicActivity.kt:
```diff
@@ -1,25 +1,33 @@
     activityComponent.inject(this)
-    topicActivityPresenter.handleOnCreate()
+    topicId = if (intent==null|| intent.getStringExtra(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) == null) TEST_TOPIC_ID_1
+      else intent.getStringExtra(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY)
```
Indentation fix in this:

Press: Option + Command + L

nikitamarysolomanpvt replied:
Resolved

rt4914 reviewed (approved):
LGTM.

@BenHenning this PR is ready for your review.

jamesxu0 commented on app/src/main/java/org/oppia/app/topic/TopicFragment.kt:
```diff
@@ -10,14 +10,19 @@ import javax.inject.Inject
-  @Inject lateinit var topicFragmentPresenter: TopicFragmentPresenter
+  @Inject
+  lateinit var topicFragmentPresenter: TopicFragmentPresenter
+  private var topicId: String = ""
```
Suggest changing variable topicId to lateinit because it is only set in onCreateView.

nikitamarysolomanpvt replied:
Resolved

jamesxu0 reviewed (approved):
LGTM

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

nikitamarysolomanpvt committed:
nit changes

BenHenning commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -55,13 +57,18 @@ class HomeFragmentPresenter @Inject constructor(
     })
   }
+
+  fun topicButtonClick(v: View) {
```
This method name should be a verb phrase, consider: ``handleTopicButtonClicked``

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicActivity.kt:
```diff
@@ -1,25 +1,32 @@
     super.onCreate(savedInstanceState)
     activityComponent.inject(this)
-    topicActivityPresenter.handleOnCreate()
+    topicId = if (intent == null || intent.getStringExtra(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) == null) TEST_TOPIC_ID_0
```
Please either assign topicId in the if/else block, or move the if/else block to a method and call it here. As it stands, this is a bit hard to parse and understand.

Alternatively, I think you can simplify this by relying on Kotlin's built-in nullness mechanism to avoid the check along with the elvis operator:

```kotlin
topicId = intent?.getStringExtra(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) ?: TEST_TOPIC_ID_0
```

nikitamarysolomanpvt replied:
This changes is already implemented in #285 as per [https://github.com/oppia/oppia-android/pull/285#discussion_r341393518](url).
I used topicId = intent?.getStringExtra(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) ?: TEST_TOPIC_ID_0

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicActivity.kt:
```diff
@@ -51,5 +58,11 @@ class TopicActivity : InjectableAppCompatActivity(), RouteToQuestionPlayerListen
 
   companion object {
     internal const val TAG_CONCEPT_CARD_DIALOG = "CONCEPT_CARD_DIALOG"
+    /** Returns a new [Intent] to route to [TopicActivity] for a specified topic ID. */
```
Nit: empty newline above this method for spacing.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragment.kt:
```diff
@@ -10,14 +10,19 @@ import javax.inject.Inject
   override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-    return topicFragmentPresenter.handleCreateView(inflater, container)
+    @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+    topicId = if (arguments != null && arguments!!.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) != null)
```
Similarly, let's use Kotlin's nullability to simplify this:

```kotlin
topicId = checkNotNull(arguments?.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY)) {
  "Expected topic ID to be included in arguments for TopicFragment."
}
```

I think this also let's us avoid needing the suppression above--please remove that.

This approach ensures that we always have a value present, and avoids needing to deal with a potentially incorrect default (empty string).

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/ViewPagerAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+import org.oppia.app.topic.review.TopicReviewFragment
+import org.oppia.app.topic.train.TopicTrainFragment
+
+class ViewPagerAdapter(fm: FragmentManager, private val numOfTabs: Int, private val topicId: String) :
```
Please add a KDoc for this class.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/ViewPagerAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+import org.oppia.app.topic.review.TopicReviewFragment
+import org.oppia.app.topic.train.TopicTrainFragment
+
+class ViewPagerAdapter(fm: FragmentManager, private val numOfTabs: Int, private val topicId: String) :
```
Prefer ``fragmentManager`` over ``fm`` to provide better context. In general, prefer avoiding abbreviated variable names.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/ViewPagerAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  override fun getCount(): Int {
+    return numOfTabs
```
We should fix this at 4, right? Let's use the enum suggested above to simplify this, e.g.:

``return TopicTab.values.size``

That way if we add a new tab this will automatically work correctly.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/ViewPagerAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+  override fun getItem(position: Int): Fragment? {
+    val args = Bundle()
+    args.putString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY, topicId)
+    when (position) {
```
Suggest using an enum to have type safety with each tab, and automatically tie them to a total count. Consider something like:

```kotlin
override fun getItem(position: Int): Fragment? {
  val args = ...
  return when (TopicTab.getTabForPosition(position)) {
    TopicTab.OVERVIEW -> {
      val topicOverviewTab = TopicOverviewFragment()
      tabOverviewTab.arguments = args
      topicOverviewTab
    }
    TopicTab.PLAY -> { ... }
    TopicTab.TRAIN -> { ... }
    TopicTab.REVIEW -> { ... }
  }
}

private enum class TopicTab(val position: Int) {
  OVERVIEW(position = 0),
  PLAY(position = 1),
  TRAIN(position = 2),
  REVIEW(position = 3)

  companion object {
    fun getTabForPosition(position: Int): TopicTab {
      val ordinal = checkNotNull(TopicTab.values.map(TopicTab::getPosition).find(position)) {
        "No tab corresponding to position: $position"
      }
      return TopicTab.values[ordinal]
    }
  }
}
```

Note that no ``else`` case needs to be handled here since all of the tabs are handled in the ``when`` statement, and the static method ensures the position integer is always correctly mapped to a tab enum constant.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/res/layout/topic_fragment.xml:
```diff
@@ -1,18 +1,48 @@
+        android:layout_height="?attr/actionBarSize"
+        android:background="?attr/colorPrimary"
+        app:popupTheme="@style/ThemeOverlay.AppCompat.Light"
+        app:title="Topic" />
```
Please use a string resource for this to enable translations for the activity title.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/res/layout/topic_fragment.xml:
```diff
@@ -1,18 +1,48 @@
+      android:layout_width="match_parent"
       android:layout_height="wrap_content"
-      android:text="This is dummy TextView for testing"
+      android:theme="@style/ThemeOverlay.AppCompat.Dark.ActionBar"
```
Why are we mixing the dark & light themes?

nikitamarysolomanpvt replied:
@BenHenning Changed it to Dark in both places

BenHenning replied:
But aren't we using a light theme at the activity level? I'm trying to understand why we're using dark here but light at the activity level. We should be consistent everywhere, otherwise switching to dark mode might have weird side effects.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -24,7 +24,11 @@
   <string name="state_learn_again_button">LEARN AGAIN</string>
   <string name="topic_overview_see_more">See More</string>
-  <string name="topic_download_text">(%s MB)</string>
+  <string name="topic_download_text"> (%s MB)</string>
```
Why this change? Please remove the extra space.

nikitamarysolomanpvt replied:
Removed

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -24,7 +24,11 @@
   <string name="topic_overview_see_more">See More</string>
-  <string name="topic_download_text">(%s MB)</string>
+  <string name="topic_download_text"> (%s MB)</string>
+  <string name="overview">OVERVIEW</string>
```
We should be capitalizing via textAllCaps via the theme instead of in the string file (see https://stackoverflow.com/questions/31295116/androidtextallcaps-false-not-working-for-tablayout-design-support for a reference, though in that they are trying to do the opposite--remove the capitalization).

nikitamarysolomanpvt replied:
suggestions where not working so used app:tabTextAppearance="@android:style/TextAppearance.DeviceDefault.Widget.TabWidget"

BenHenning commented on app/src/sharedTest/java/org/oppia/app/utility/EspressoTestsMatchers.kt:
```diff
@@ -2,11 +2,15 @@ package org.oppia.app.utility
 
+  fun matchCurrentTabTitle(tabTitle: String): Matcher<View> {
+    return matchCurrentTabTitleCheck(tabTitle)
+  }
```
Nit: blank newline after this method for spacing.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/utility/TabMatcher.kt:
```diff
@@ -0,0 +1,53 @@
+import com.google.android.material.tabs.TabLayout
+import org.hamcrest.Matcher
+
+
```
Nit: please remove extra unneeded blank newline.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/utility/TabMatcher.kt:
```diff
@@ -0,0 +1,53 @@
+      }
+    }
+
+    // Currently its not getting used
```
Do we anticipate this to be used? If not, please remove it.

nikitamarysolomanpvt replied:
Removed

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+    TopicActivity::class.java, /* initialTouchMode= */ true, /* launchActivity= */ false
+  )
+
+  private fun getResourceString(id: Int): String {
```
Nit: please move helpers to be after the list of tests.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+
+  @Test
+  fun testTopicFragment_showsTopicFragmentWithMultipleTabs() {
+    activityTestRule.launchActivity(null)
```
Why is this needed? The rule should automatically launch the activity.

nikitamarysolomanpvt replied:
Removed

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+  fun testTopicFragment_swipePage_hasSwipedPage() {
+    activityTestRule.launchActivity(null)
+    onView(withId(R.id.topic_tabs_viewpager)).check(matches(isDisplayed()))
+    onView(withId(R.id.topic_tabs_viewpager)).perform(ViewActions.swipeLeft())
```
Will this fail? There doesn't appear to be any verification happening after the isDisplayed() above.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+  }
+
+  @Test
+  @UiThread
```
Why is this needed? From what I understand ``@UiThread`` is just a note, not actually an enforced annotation (e.g. adding it will not ever change functionality of code it's annotating). Please remove it if it isn't needed.

Ditto below.

nikitamarysolomanpvt replied:
Removed

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+
+  @Test
+  @UiThread
+  fun testTopicFragment_overviewTopicIsDisplayedInTabLayout_worksAsExpected() {
```
What's being verified here? Is the descendent check the key thing? If so, we should emphasize that's being checked:

```kotlin
fun testTopicFragment_overviewTopicTab_isDisplayedInTabLayout() {
  onView(withText(R.string.overview)).check(matches(isDescendantOfA(withId(R.id.topic_tabs_container)))
}
```

Ditto for other similar tests below.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+      allOf(
+        atPosition(R.id.review_skill_recycler_view, 0).also { withText("An important skill") },
+        isDescendantOfA(withId(R.id.review_skill_recycler_view))
+      ).also { withId(R.id.skill_name) })
```
Here & above: I believe this ``also`` isn't actually doing anything since it's a Kotlin scope function: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/also.html.

Shouldn't this be a check(matches()) case?

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+
+  @Test
+  @UiThread
+  fun testTopicFragment_clickOnOverviewTab_showsOverviewTabWithContentMatched() {
```
Is this testing tab switching? If so, please indicate in the test name or remove the extra switching in the test, since the two don't match.

E.g. the test name could be: ``testTopicFragment_clickOnReviewTab_thenOverviewTab_showsOverviewTab``

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+      )
+    ).perform(click())
+    onView(withText("First Story")).check(matches(isDisplayed()))
+    activityTestRule.activity.requestedOrientation = Configuration.ORIENTATION_LANDSCAPE
```
Doesn't this need to happen in an ``onActivity()`` callback? As I recall from documentation, requestedOrientation should only be called on the main thread. Ditto elsewhere.

nikitamarysolomanpvt replied:
Is this the solution you are suggesting ,Please take a look on this link [https://github.com/oppia/oppia-android/pull/289#discussion_r342008956 ](url). 
It seems not working.
Please do suggest the link if we have any google documentation for this fix.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+    onView(
+      allOf(
+        withText(R.string.train),
+        isDisplayed()
```
This should be a check(matches()) instead of being embedded in the onView(). Ditto below.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,350 @@
+      )
+    ).perform(click())
+    onView(allOf(
+      atPosition(R.id.review_skill_recycler_view, 0).also { withText("An important skill") },
```
Please fix this ``also`` cases here & elsewhere in this test suite.

nikitamarysolomanpvt replied:
Resolved

BenHenning reviewed (changes requested):
Thanks @nikitamarysolomanpvt! I mostly had nits, though I had various concerns with the new test suite. Would like to do one more pass given those.

nikitamarysolomanpvt committed:
nit changes

rt4914 commented:
@nikitamarysolomanpvt The description of #135 has been updated and based on that please change the title of this PR to `Fix part of #135 `

nikitamarysolomanpvt committed:
set title on the basis of selected topic in topic screen.other nit changes

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'topic-player-multiple-tabs' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

# Conflicts:
#	app/src/main/java/org/oppia/app/home/HomeActivity.kt
#	app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt
#	app/src/main/java/org/oppia/app/topic/TopicActivity.kt
#	app/src/main/res/layout/home_fragment.xml

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'topic-player-multiple-tabs' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'topic-player-multiple-tabs' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

# Conflicts:
#	app/src/main/java/org/oppia/app/home/HomeActivity.kt
#	app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt
#	app/src/main/java/org/oppia/app/topic/TopicActivity.kt
#	app/src/main/res/layout/home_fragment.xml

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'topic-player-multiple-tabs' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

# Conflicts:
#	app/src/main/AndroidManifest.xml

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'topic-player-multiple-tabs' of https://github.com/oppia/oppia-android into topic-player-multiple-tabs

# Conflicts:
#	app/src/main/AndroidManifest.xml

veena14cs commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -3,21 +3,88 @@ package org.oppia.app.topic
+
+  private fun subscribeToTopicLiveData() {
+    topicLiveData.observe(fragment, Observer<Topic> { result ->
+      var topicName = result.name
```
This can be declared as val.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt committed:
nit in TopicFragmentPresenter changed var to val in line number 69

veena14cs reviewed (approved):
LGTM

nikitamarysolomanpvt committed:
shorten stackoverflow link,testcase method name change

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicFragment.kt:
```diff
@@ -6,18 +6,24 @@ import android.view.LayoutInflater
 
   override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-    return topicFragmentPresenter.handleCreateView(inflater, container)
+    topicId = checkNotNull(arguments?.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY)?: TEST_TOPIC_ID_0) {
```
Small Nit: `) ?: TEST_TOPIC_ID_0)`
spacing issue.

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -3,21 +3,88 @@ package org.oppia.app.topic
+    topicController.getTopic(TEST_TOPIC_ID_0)
+  }
+
+  private fun getTopicList(): LiveData<Topic> {
```
It should be `getTopic` rather than `getTopicList`

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -3,21 +3,88 @@ package org.oppia.app.topic
+
+  private fun processTopicResult(topic: AsyncResult<Topic>): Topic {
+    if (topic.isFailure()) {
+      logger.e("TopicOverviewFragment", "Failed to retrieve topic", topic.getErrorOrNull()!!)
```
TAG in this logger is incorrect. `TopicFragment`

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicTab.kt:
```diff
@@ -0,0 +1,23 @@
+
+  init {
+    this.tabPosition = position
+  }
```
`init` and `tabPosition` can be merged to `var tabPosition: Int = position`

Suggested by Android Studio.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicTab.kt:
```diff
@@ -0,0 +1,23 @@
+
+  companion object {
+    fun getTabForPosition(position: Int): TopicTab {
+      val ordinal = checkNotNull(TopicTab.values().map(TopicTab::tabPosition).get(position)) {
```
can be updated to `checkNotNull(values().map(TopicTab::tabPosition)[position]) `

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/main/java/org/oppia/app/topic/overview/TopicOverviewFragmentPresenter.kt:
```diff
@@ -61,7 +61,7 @@ class TopicOverviewFragmentPresenter @Inject constructor(
   // TODO(#135): Get this topic-id from [TopicFragment].
   private val topicResultLiveData: LiveData<AsyncResult<Topic>> by lazy {
-    topicController.getTopic(TEST_TOPIC_ID_1)
+    topicController.getTopic(TEST_TOPIC_ID_0)
```
@BenHenning This dummy id has been replaced in all test cases in #285

rt4914 commented on app/src/main/res/drawable/ic_overview_white_24dp.xml:
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+  android:width="24dp"
+  android:height="24dp"
```
There is this inconsistency in icon names and sizes.
1. width and height in all these 4 icons should either be 48dp each or 24dp each.
2. name should either end with `_24dp.xml` or not but it should be same for all.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

rt4914 replied:
These file names are still not consistent, 

every file name should end with `_24dp`

nikitamarysolomanpvt replied:
Done

rt4914 commented on app/src/main/res/values/styles.xml:
```diff
@@ -70,4 +78,6 @@
     <item name="android:clickable">false</item>
   </style>
+
+
```
Remove this extra blank space.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

rt4914 commented on app/src/sharedTest/java/org/oppia/app/utility/TabMatcher.kt:
```diff
@@ -0,0 +1,32 @@
+import org.hamcrest.Matcher
+
+/**
+ * @see <a href="https://stackoverflow.com/questions/49626315">TabMatcher</a>
```
`Optional Nit`: This is just for better information to @nikitamarysolomanpvt . You can make this link even shorter by using `https://stackoverflow.com/q/49626315`, i.e, `q` in place of `questions`.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

rt4914 reviewed (approved):
This looks like a very clean implementation to me. Great work.

There are some nit changes which you can fix.

nikitamarysolomanpvt committed:
nit changes suggested by Rajat

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -3,21 +3,88 @@ package org.oppia.app.topic
+
+  private fun processTopicResult(topic: AsyncResult<Topic>): Topic {
+    if (topic.isFailure()) {
+      logger.e("TopicFragmentPresenter", "Failed to retrieve topic", topic.getErrorOrNull()!!)
```
Nit: Logger should reference to fragment or activity only and not presenter or viewmodel

nikitamarysolomanpvt replied:
Resolved

rt4914 reviewed (approved):
LGTM

rt4914 commented:
@nikitamarysolomanpvt Because of latest code merge your PR now has merge conflicts, please resolve them.

nikitamarysolomanpvt committed:
merged with develop

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicActivity.kt:
```diff
@@ -9,19 +9,25 @@ import org.oppia.app.player.exploration.ExplorationActivity
     activityComponent.inject(this)
-    topicActivityPresenter.handleOnCreate()
+    topicId  = checkNotNull(intent?.getStringExtra(org.oppia.app.topic.TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY)?: TEST_TOPIC_ID_0) {
+      "Expected topic ID to be included in arguments for TopicFragment."
```
As this code stands, I don't think it will ever throw. I think you need to remove the elvis operator if you want it to fail when there's no result (at the moment it will default to TEST_TOPIC_ID_0.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt committed:
nit

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragment.kt:
```diff
@@ -6,18 +6,24 @@ import android.view.LayoutInflater
 
   override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-    return topicFragmentPresenter.handleCreateView(inflater, container)
+    topicId = checkNotNull(arguments?.getString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY) ?: TEST_TOPIC_ID_0) {
```
Ditto here.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -3,21 +3,88 @@ package org.oppia.app.topic
+    })
+  }
+
+  // TODO(#135): Get this topic-id from [TopicFragment].
```
Can we resolve this now? It seems like the topic ID is being passed in from ``TopicFragment`` per other changes in this PR.

rt4914 replied:
This is done in next PR #285

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicFragmentPresenter.kt:
```diff
@@ -3,21 +3,88 @@ package org.oppia.app.topic
+  fun handleCreateView(
+    inflater: LayoutInflater,
+    container: ViewGroup?,
+    topicId: String?
```
Why is this nullable? We are ensuring it isn't in ``TopicFragment``, so I think we can rely on it not being nullable.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicTab.kt:
```diff
@@ -0,0 +1,19 @@
+package org.oppia.app.topic
+
+enum class TopicTab(val position: Int) {
```
Please add a KDoc for this enum to provide context on what it represents.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicTab.kt:
```diff
@@ -0,0 +1,19 @@
+package org.oppia.app.topic
+
+enum class TopicTab(val position: Int) {
```
Can ``position`` be private? Prefer minimizing visibility.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/main/java/org/oppia/app/topic/TopicTab.kt:
```diff
@@ -0,0 +1,19 @@
+  TRAIN(position = 2),
+  REVIEW(position = 3);
+
+  var tabPosition: Int = position
```
Why do we need this? ``position`` already contains the value, and this shouldn't ever change (so it shouldn't be a ``var``).

nikitamarysolomanpvt replied:
Removed

BenHenning commented on app/src/main/java/org/oppia/app/topic/ViewPagerAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+
+  override fun getItem(position: Int): Fragment? {
+    val args = Bundle()
+    args.putString(TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY, topicId)
```
I feel like we should use a key defined in this adapter rather than the activity one. We ought to have separate keys for arguments and intent values for each activity/fragment, but in this case I think it's reasonable to have a shared key for each of the tab fragments that's defined here to simplify things.

nikitamarysolomanpvt replied:
Renamed when sent to fragment.

nikitamarysolomanpvt commented:
> @nikitamarysolomanpvt Because of latest code merge your PR now has merge conflicts, please resolve them.

Merged with develop

BenHenning commented on app/src/main/res/layout/home_fragment.xml:
```diff
@@ -16,19 +16,35 @@
+        android:layout_height="wrap_content"
+        android:layout_marginTop="16dp"
+        android:onClick="@{presenter::handleTopicButtonClicked}"
+        android:text="Go to topic" />
```
Why do we need this? We should load topics based on the recycler view. Let's not add anymore temporary buttons in the home screen--we need to rely on existing UIs at this point.

rt4914 replied:
Considering this, we might need to change the testing strategy meaning, creating a new TestActivity which loads TopicFragment for a specific TOPIC_ID.

BenHenning replied:
Yes, that's what we should be doing. That also future proofs this as we start loading real data in the domain layer since we can always load a test topic ID for the purpose of these tests.

BenHenning replied:
@nikitamarysolomanpvt since this requires downstream test changes, I think it's fine to do this in a later PR as part of resolving #135 if it isn't straightforward.

nikitamarysolomanpvt replied:
Okay will create separate PR for this

BenHenning commented on app/src/main/res/layout/topic_fragment.xml:
```diff
@@ -1,18 +1,48 @@
-      android:layout_width="wrap_content"
+
+    <com.google.android.material.appbar.AppBarLayout
+      android:id="@+id/appBarLayout"
```
Layout IDs should be defined using ``snake_case``. Also, let's make this contextualized based on the container since Android IDs are global, and 'appBarLayout' isn't clear which layout it corresponds to when referenced in Kotlin.

Suggest: ``topic_app_bar_layout``.

nikitamarysolomanpvt replied:
resolved

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -25,6 +25,11 @@
   <string name="state_learn_again_button">LEARN AGAIN</string>
   <string name="topic_overview_see_more">See More</string>
   <string name="topic_download_text">(%s MB)</string>
+  <string name="overview">Overview</string>
```
String names should have context or they won't be clear when referenced in code. Suggest: ``topic_tab_overview``, ``topic_tab_play``, etc.

nikitamarysolomanpvt replied:
Removed

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -25,6 +25,11 @@
+  <string name="play">Play</string>
+  <string name="train">Train</string>
+  <string name="review">Review</string>
+  <string name="topic">Topic</string>
```
This string appears to be unused--please remove.

nikitamarysolomanpvt replied:
Removed

BenHenning commented on app/src/sharedTest/java/org/oppia/app/utility/TabMatcher.kt:
```diff
@@ -0,0 +1,32 @@
+
+/**
+ * @see <a href="https://stackoverflow.com/q/49626315">TabMatcher</a>
+ * This class mainly provides a custom matcher to test whether the current tab title is correctly selected in TabLayout. */
```
We should format KDocs consistently, and it's helpful to list the purpose of the class before the reference. Suggest:

```kotlin
/**
 * A custom matcher to test whether the current tab title is correctly selected in a [TabLayout].
 *
 * Reference: https://stackoverflow.com/q/49626315
 */
```

(Note that KDoc uses Markdown rather than HTML, but in this case I think we can just reference the URL directly rather than naming it ).

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/utility/TabMatcher.kt:
```diff
@@ -0,0 +1,32 @@
+          val tabLayout = item as TabLayout
+          val tabAtIndex: TabLayout.Tab = tabLayout.getTabAt(tabLayout.selectedTabPosition)
+            ?: throw PerformException.Builder()
+              .withCause(Throwable("No tab at index ${tabLayout.selectedTabPosition}"))
```
Let's use IllegalStateException instead of Throwable. Prefer not to throw top-level types like Throwable, Exception, or Error.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,342 @@
+
+  private fun getResourceString(id: Int): String {
+    val resources =
+      InstrumentationRegistry.getInstrumentation().targetContext.resources
```
I believe this won't work correctly when run on Robolectric, and I think we can rely on the AndroidX application provider to get access to resources:

```kotlin
val resources = ApplicationProvider.getApplicationContext().resources
```

nikitamarysolomanpvt replied:
This is giving an error 
`/Users/marysoloman/StudioProjects/oppia-android/app/build/tmp/kapt3/stubs/debug/org/oppia/app/application/ApplicationComponent.java:11: error: [Dagger/MissingBinding] org.oppia.app.home.UserAppHistoryViewModel cannot be provided without an @Inject constructor or an @Provides-annotated method.
public abstract interface ApplicationComponent {
                ^
      javax.inject.Provider<org.oppia.app.home.UserAppHistoryViewModel> is injected at
          org.oppia.app.viewmodel.ViewModelBridgeFactory(viewModelProvider)
      org.oppia.app.viewmodel.ViewModelBridgeFactory<org.oppia.app.home.UserAppHistoryViewModel> is injected at
          org.oppia.app.viewmodel.ViewModelProvider(bridgeFactory)
      org.oppia.app.viewmodel.ViewModelProvider<org.oppia.app.home.UserAppHistoryViewModel> is injected at
          org.oppia.app.home.HomeFragmentPresenter(…, viewModelProvider, …)
      org.oppia.app.home.HomeFragmentPresenter is injected at
          org.oppia.app.home.HomeFragment.homeFragmentPresenter
      org.oppia.app.home.HomeFragment is injected at
          org.oppia.app.fragment.FragmentComponent.inject(org.oppia.app.home.HomeFragment) [org.oppia.app.application.ApplicationComponent → org.oppia.app.activity.ActivityComponent → org.oppia.app.fragment.FragmentComponent]
`

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,342 @@
+import org.oppia.util.threading.BlockingDispatcher
+import javax.inject.Singleton
+
+/** Tests for [TopicActivity]. */
```
Nit: ``Tests for [TopicFragment]``

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,342 @@
+/** Tests for [TopicActivity]. */
+@RunWith(AndroidJUnit4::class)
+class TopicFragmentTest {
+  private val topicName = "First Test Topic"
```
Let's make this a file-level constant instead of a property, e.g.:

```kotlin
private const val FIRST_TOPIC_NAME = "First Test Topic"

@RunWith(AndroidJUnit4::class)
class TopicFragmentTest {
  ...
}
```

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt replied:
Resolved

BenHenning reviewed (changes requested):
Thanks @nikitamarysolomanpvt--really solid tests!

BenHenning reviewed (approved):
Thanks @nikitamarysolomanpvt -- the PR generally LGTM, and I'm deferring on @rt4914 to verify that the specific concerns I brought up in my comments are addressed. Approving so that this can be checked in later today.

nikitamarysolomanpvt commented:
> Thanks @nikitamarysolomanpvt -- the PR generally LGTM, and I'm deferring on @rt4914 to verify that the specific concerns I brought up in my comments are addressed. Approving so that this can be checked in later today.

I would like to make the new suggested changes in the #285. Because #285 is dependent on #230 as well as it is base branch of #285. If you want i can create issue on the same.

nikitamarysolomanpvt committed:
changes suggested by ben https://github.com/oppia/oppia-android/pull/230#pullrequestreview-312215773

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicActivity.kt:
```diff
@@ -9,19 +9,23 @@ import org.oppia.app.player.exploration.ExplorationActivity
     super.onCreate(savedInstanceState)
     activityComponent.inject(this)
-    topicActivityPresenter.handleOnCreate()
+    topicId  = intent?.getStringExtra(org.oppia.app.topic.TOPIC_ACTIVITY_TOPIC_ID_ARGUMENT_KEY)?: TEST_TOPIC_ID_0
```
Nit: Remove one extra space

`topicId = `

nikitamarysolomanpvt replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicFragment.kt:
```diff
@@ -6,18 +6,22 @@ import android.view.LayoutInflater
 
   override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-    return topicFragmentPresenter.handleCreateView(inflater, container)
+    topicId = arguments?.getString(TOPIC_ID_ARGUMENT_KEY) ?: TEST_TOPIC_ID_0
```
**NOTE**: `TEST_TOPIC_ID_0` will be removed in either #285 or a separate PR. For now we need this successfully running the test-cases.

rt4914 commented on app/src/main/java/org/oppia/app/topic/ViewPagerAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+    return TopicTab.values().size
+  }
+}
+
```
Remove this extra blank line.

nikitamarysolomanpvt replied:
Done

rt4914 commented on app/src/sharedTest/java/org/oppia/app/topic/TopicFragmentTest.kt:
```diff
@@ -0,0 +1,336 @@
+import org.oppia.util.threading.BlockingDispatcher
+import javax.inject.Singleton
+
+private val topicName = "First Test Topic"
```
Update this to `private const val TOPIC_NAME = "First Test Topic"`

nikitamarysolomanpvt replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/topic/TopicTab.kt:
```diff
@@ -0,0 +1,18 @@
+package org.oppia.app.topic
+
+/** Enum to store the tabs of topic and get tab by position. */
```
Update this comment to 
`/** Enum to store the tabs of [TopicFragment] and get tab by position. */`

Topic is a very general thing which might reference to the topic coming from oppia-backend, domain module, etc. but these tabs are actually specifically related to `TopicFragment file`.

nikitamarysolomanpvt replied:
Used the suggested comment

rt4914 commented on app/src/sharedTest/java/org/oppia/app/utility/TabMatcher.kt:
```diff
@@ -0,0 +1,34 @@
+          val tabLayout = item as TabLayout
+          val tabAtIndex: TabLayout.Tab = tabLayout.getTabAt(tabLayout.selectedTabPosition)
+            ?: throw PerformException.Builder()
+              .withCause(throw IllegalStateException("No tab at index ${tabLayout.selectedTabPosition}"))
```
This code is giving `Unreachable code` warning can you fix that.

nikitamarysolomanpvt replied:
throw IllegalStateException
Suggested by ben

nikitamarysolomanpvt replied:
Resolved

rt4914 reviewed (approved):
LGTM, there are still nit change which needs to be fixed.

Also, I suggest before merging this PR mention one comment regarding what all work which was not finished in this PR, so that anyone going through this PR in future has clear context of our verbal conversations.

Sample:
 - Removing dummy button from home and linking `TopicActivity` to `Topic` items of `HomeFragment`.
 - Introducing `TopicTestActivity` and its corresponding presenter to load `TopicFragment` for a specific TOPIC_ID which will be used to test-code.
 - Transfer of TOPIC_ID from TopicFragment to ChildFragment is pending
 - If the TOPIC_ID and STORY_ID is present in `TopicActivity` then the UI should traverse to a specific STORY inside `TopicPlayFragment`

If you think something else is missing feel free to add that too.

nikitamarysolomanpvt committed:
nit

veena14cs reviewed (approved):
LGTM.

rt4914 commented:
> LGTM, there are still nit change which needs to be fixed.
> 
> Also, I suggest before merging this PR mention one comment regarding what all work which was not finished in this PR, so that anyone going through this PR in future has clear context of our verbal conversations.
> 
> Sample:
> 
> * Removing dummy button from home and linking `TopicActivity` to `Topic` items of `HomeFragment`.
> * Introducing `TopicTestActivity` and its corresponding presenter to load `TopicFragment` for a specific TOPIC_ID which will be used to test-code.
> * Transfer of TOPIC_ID from TopicFragment to ChildFragment is pending
> * If the TOPIC_ID and STORY_ID is present in `TopicActivity` then the UI should traverse to a specific STORY inside `TopicPlayFragment`
> 
> If you think something else is missing feel free to add that too.

@nikitamarysolomanpvt I think you missed out on this comment before merging the PR.

nikitamarysolomanpvt commented:
> Also, I suggest before merging this PR mention one comment regarding what all work which was not finished in this PR, so that anyone going through this PR in future has clear context of our verbal conversations.
> 
> Sample:
> 
> * Removing dummy button from home and linking `TopicActivity` to `Topic` items of `HomeFragment`.
> * Introducing `TopicTestActivity` and its corresponding presenter to load `TopicFragment` for a specific TOPIC_ID which will be used to test-code.
> * Transfer of TOPIC_ID from TopicFragment to ChildFragment is pending
> * If the TOPIC_ID and STORY_ID is present in `TopicActivity` then the UI should traverse to a specific STORY inside `TopicPlayFragment` 
> 
> If you think something else is missing feel free to add that too.

Sorry i missed this comment.
Adding a few more
 * If the TOPIC_ID and STORY_ID is present in `TopicActivity` then the UI should traverse to a specific STORY inside `TopicPlayFragment` and expand the item.
 * If the TOPIC_ID and STORY_ID is present in `TopicActivity` then the UI should traverse to a specific STORY inside `TopicPlayFragment` (resolved in #285).
